crates/telemetry/src/detector.rs

```diff
@@ -0,0 +1,49 @@
+use std::{env, time::Duration};
+
+use opentelemetry::{Key, KeyValue, Value};
+use opentelemetry_sdk::{
+    resource::{EnvResourceDetector, ResourceDetector},
+    Resource,
+};
+
+const OTEL_SERVICE_NAME: &str = "OTEL_SERVICE_NAME";
+
+/// Custom resource detector for Spin relevant attributes service.name and service.version.
+///
+/// To set service.name this detector will first try `OTEL_SERVICE_NAME` env. If it's not available,
+/// then it will check the `OTEL_RESOURCE_ATTRIBUTES` env and see if it contains `service.name`
+/// resource. If it's also not available, it will use `spin`.
+///
+/// To set service.version, it will use the version passed in new.
+#[derive(Debug)]
+pub struct SpinResourceDetector {
+    version: String,
+}
+
+impl SpinResourceDetector {
+    /// Create a new instance of SpinResourceDetector.
+    pub fn new(version: String) -> Self {
+        SpinResourceDetector { version }
+    }
+}
+
+impl ResourceDetector for SpinResourceDetector {
+    fn detect(&self, _timeout: Duration) -> Resource {
+        Resource::new(vec![
+            KeyValue::new(
+                "service.name",
+                env::var(OTEL_SERVICE_NAME)
```

Comment:
This feels like a good candidate for being extracted to a variable and given a meaningful name.